talk/expert/cpp20spaceship.tex

```diff
@@ -0,0 +1,250 @@
+     {
+      unsigned n, d ;
+
+      friend std::weak_ordering operator<=>( Rational const & a, Rational const & b )
+       { return (a.n*b.d)<=>(a.n*b.d) ; }
+     } ;
+
+    std::ostream & operator<<( std::ostream & os, std::weak_ordering cmp )
+     { return (os<<'<'<<(cmp<0)<<'|'<<(cmp==0)<<'|'<<(cmp>0)<<'>') ; }
+
+    std::ostream & operator<<( std::ostream & os, Rational const & r )
+     { return (os<<r.n<<'/'<<r.d) ; }
+
+    template <typename T>
+    void compare( T lhs, T rhs )
+     {
+      std::cout<<std::endl ;
+      //std::cout<<lhs<<" ==  "<<rhs<<": "<<(lhs==rhs)<<std::endl ;
+      //std::cout<<lhs<<" !=  "<<rhs<<": "<<(lhs!=rhs)<<std::endl ;
+      std::cout<<lhs<<" <=> "<<rhs<<": "<<(lhs<=>rhs)<<std::endl ;
+      std::cout<<lhs<<" <   "<<rhs<<": "<<(lhs<  rhs)<<std::endl ;
+      std::cout<<lhs<<" <=  "<<rhs<<": "<<(lhs<= rhs)<<std::endl ;
+      std::cout<<lhs<<" >   "<<rhs<<": "<<(lhs>  rhs)<<std::endl ;
+      std::cout<<lhs<<" >=  "<<rhs<<": "<<(lhs>= rhs)<<std::endl ;
+     }
+
+    int main()
+     {
+      compare<Rational>({ 3, 4 },{ 2, 3 }) ;
+      compare<Rational>({ 3, 6 },{ 2, 3 }) ;
+      compare<Rational>({ 1, 2 },{ 2, 4 }) ;
+     }
+     \end{cppcode*}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{The three-way comparison operator}
+  \begin{block}{Home-made class}
+    \begin{itemize}
+      \item Despites \mintinline{cpp}{<=>} for \mintinline{cpp}{unsigned} returning a \mintinline{cpp}{std::strong_ordering}, we prefer here to convert it to \mintinline{cpp}{std::weak_ordering}. This way, we emphasize that if \mintinline{cpp}{a<=>b} is equal to \mintinline{cpp}{0}, it only means that \mintinline{cpp}{a} and \mintinline{cpp}{b} are logically equivalent, but may lead to different results in other expressions (e.g. printing them).
+      \item As one can see from the code, we have defined only \mintinline{cpp}{<=>}, but \mintinline{cpp}{<}, \mintinline{cpp}{>}, \mintinline{cpp}{<=} and \mintinline{cpp}{=>} work as well. The compiler can express the latter operators in terms of \mintinline{cpp}{<=>}. Of course, one can also provide own implementations.
+      \item It has NOT added a default implementation for \mintinline{cpp}{==} and \mintinline{cpp}{!=}. Those operators are generally expected to mean \textbf{equal}, rather than \textbf{equivalent}. If \mintinline{cpp}{<=>} does not provides a strong order, it is generally advised not to define \mintinline{cpp}{==}.
+    \end{itemize}
+  \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{The three-way comparison operator}
+  \begin{block}{Default \mintinline{cpp}{<=>} implementation}
+    \begin{itemize}
+      \item One can ask the compiler to provide a default implementation for \mintinline{cpp}{<=>} and/or  \mintinline{cpp}{==}. Logically enough, it will compare the first member variable of the two objects, and goes on to the next member variable as long as the current ones are equivalent.
```

Comment:
```suggestion
      \item One can ask the compiler to provide a default implementation for \mintinline{cpp}{<=>} and/or \mintinline{cpp}{==}. Logically enough, it will compare the first member variable of the two objects, and goes on to the next member variable as long as the current ones are equivalent.
```